File: tests/omp/tasking13.ml
open Optitrust
open Target

let _ = Run.script_cpp (fun _ ->
  
  !! Omp.task [Final "pos > LIMIT"; Mergeable] [tBefore; tIndex ~nb:3 0; cIf ()];
  !! Omp.task [Final "pos > LIMIT";Mergeable] [tAfter; tIndex ~nb:2 0; cSeq ~args_pred:(Target.target_list_one_st [cVarDef "new_state"]) ()];
  !! Omp.taskwait [tAfter; tIndex ~nb:2 1; cSeq ~args_pred:(Target.target_list_one_st [cVarDef "new_state"]) ()];
)

File: tests/omp/tasking13.cpp
#include <string.h>
#include <omp.h>
#include <stdio.h>

const int LIMIT = 3; /* arbitrary limit on recursion depth */

void check_solution(char *p);
void bin_search (int pos, int n, char *state)
{
   if ( pos == n ) {
      check_solution(state);
      return;
   }
   {
      char new_state[n];
      if (!omp_in_final() ) {
        memcpy(new_state, state, pos );
        state = new_state;
      }
      state[pos] = 0;
      bin_search(pos+1, n, state );
   }
   {
      char new_state[n];
      if (! omp_in_final() ) {
        memcpy(new_state, state, pos );
        state = new_state;
      }
      state[pos] = 1;
      bin_search(pos+1, n, state );
   }
}
